src/react/provider.js

```diff
@@ -1,16 +1,42 @@
-import React, { useContext, useMemo, forwardRef, useEffect } from 'react';
+import React, { forwardRef, Component } from 'react';
 import PropTypes from 'prop-types';
 
 import { diRegistry } from './constants';
 import { Context } from './context';
 import { addInjectableToMap, getDisplayName, findInjectable } from './utils';
 import { globalDi } from './global';
 
-export const DiProvider = ({ children, use, target, global }) => {
-  const { getDependencies } = useContext(Context);
+export class DiProvider extends Component {
+  static contextType = Context;
+
+  static propTypes = {
+    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
+    global: PropTypes.bool,
+    target: PropTypes.oneOfType([
+      PropTypes.func,
+      PropTypes.arrayOf(PropTypes.func),
+    ]),
+    use: PropTypes.arrayOf(
+      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
+    ).isRequired,
+  };
+
+  componentDidCatch(err) {
+    globalDi._remove(this.props.use);
+    throw err;
+  }
+
+  componentWillUnmount() {
+    globalDi._remove(this.props.use);
```

Comment:
these are not the same `use` you've got during the initial render and later used for `_fromProvider`, these are different.

There were a few bugs due to mismatches of this sort, I would better remeber a "cleanup callback" somewhere in `getValue`

Reply:
How come? I would expect them to be consistent 🤔 Unless someone decides to conditionally add/remove injectables (which we call out as not supported) it should contain the same "stuff"

Reply:
You know, stuff happens.
This is not a real problem, but a potential.

Ideally just expose another (secret) method to check if global DI is clear. Then we can call it before test to verify that everything is as it should be. Later we can remove this method.

Well, it should be fine, but we all know that our assumptions are not always true